src/listfunc.c

```diff
@@ -1524,6 +1527,112 @@ static Obj FuncLIST_WITH_IDENTICAL_ENTRIES(Obj self, Obj n, Obj obj)
     return list;
 }
 
+
+static Obj FastCallFuncList(Obj func, Obj list)
+{
+    switch (LEN_PLIST(list)) {
+    case 0:
+        return CALL_0ARGS(func);
+    case 1:
+        return CALL_1ARGS(func, ELM_PLIST(list, 1));
+    case 2:
+        return CALL_2ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2));
+    case 3:
+        return CALL_3ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3));
+    case 4:
+        return CALL_4ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3), ELM_PLIST(list, 4));
+    case 5:
+        return CALL_5ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3), ELM_PLIST(list, 4),
+                          ELM_PLIST(list, 5));
+    case 6:
+        return CALL_6ARGS(func, ELM_PLIST(list, 1), ELM_PLIST(list, 2),
+                          ELM_PLIST(list, 3), ELM_PLIST(list, 4),
+                          ELM_PLIST(list, 5), ELM_PLIST(list, 6));
+    default:
+        return CALL_XARGS(func, list);
+    }
+}
+
+// TODO: document this a bit
+static int FoldLeftXHelp(Obj   gens,
+                         Obj   foldFunc,
+                         Obj * acc,
+                         Obj   abortValue,
+                         Obj   args,
+                         int   genIndex,
+                         int   valIndex)
+{
+    while (genIndex <= LEN_PLIST(gens)) {
+        Obj gen = ELM_PLIST(gens, genIndex);
+        if (IS_FUNC(gen))
+            gen = FastCallFuncList(gen, args);
```

Comment:
TODO: replace `FastCallFuncList ` by just `CallFuncList`? It was kind of an attempt to bypass call overhead, but now I wonder if that wasn't a premature optimization. Doing some timing measurements probably would be a good idea: comparing performance with ` `FastCallFuncList ` versus `CallFuncList`, but also that of `ListX`, `SumX`, `ProductX` before and after this PR.